plugin/completion.py

```diff
@@ -191,6 +193,19 @@ def on_query_completions(self, prefix: str, locations: List[int]) -> Optional[su
             lambda res: self.handle_error(res, completion_list))
         return completion_list
 
+    def normalized_documentation(self, lsp_item: Dict[str, Any]) -> str:
+        lsp_documentation = lsp_item.get("documentation")
```

Comment:
Is this function expensive? 
I tried the PR locally,  
When I trigger the AC in LSP-elm it takes >20s for this line finish. The AC doesn't show up at the end.
```
items = list(map(self.format_completion, response_items))
```
maybe the reason it is the combination of `normalized_documentation` and `transform_region` being applied to all items?

![output](https://user-images.githubusercontent.com/22029477/80403905-add90200-88c0-11ea-88bf-13b30842c21e.gif)

Reply:
transform_region is not that expensive, 
but normalized_documentation is. If I comment out the `normalized_documentation` calls, the AC show up

Reply:
These are some good points. Waiting potentially 20 seconds before completions show up is of course unacceptable. Performance from mdpopups.md2html is apparently not good enough (perhaps @facelessuser is willing to improve it, if that it even possible). I think the problem is ultimately on our side since we're rendering potentially >200 markdown strings to html.

Ideally, documentation would only be rendered once the user highlights a particular completion item (lazy rendering). But right now we're forced to render *all* markdown docs to html before presenting completions (cc @jskinner)

That said, from some initial experience with this PR I think putting the documentation into the ST-details is not a good fit. This is because we don't know how many lines/sentences a server will put into the docs. And there is only place for *at most* one sentence in the ST-details field.

So either we'll have to come up with a way to extract the "most important sentence" from the returned markdown string, or we have to think of something else.

One idea that's been floating in my head is as follows:

Use the new `subl://` scheme in the ST-details to make a href to a command (let's call it `LspShowDocumentationCommand`). So that href will appear in the ST-details field at the bottom of the completion widget. The user can click it. The `LspShowDocumentationCommand` would then show the docs in a popup with `mdpopups.show_popup` and with the flag `sublime.COOPERATE_WITH_AUTO_COMPLETE`. This would allow us to do the lazy rendering as we can store the LSP-documentations of all items somewhere.

Reply:
Because we currently don't know the what the highlighted completion item is in the AC, 
Just to chain on your idea. 
Instead of just calling LspShowDocumentationCommand, 
the documentation field may not be there unless that completion item is resolved.
I would just suggest first calling the `completionItem/resolve` before showing the documentation.

Reply:
It is the same for Metals, client is expected to call `completionItem/resolve` while scrolling through the completion list because the documentation is expensive to compute. 
Is it possible to request `completionItem/resolve` via the new AC API ?

Reply:
We can fetch docs by making a hyperlink to a command in the ST-details field: `<a href="subl:some_command">Docs...</a>`

But the user will have to click on it with the mouse, or use a keybinding. I want to implement this in a next PR.

Reply:
Do we know up front if **there is** some documentation available for item?
It would be a bad experience to present a link for all items just to return nothing after clicking it.

Reply:
@rchl I believe LSP knows if the server supports `completionItem/resolve`. So no link will be shown if the sever can't provide documentation.

@rwols why forcing the user to click ? I was under the impression we can know show a popup along side the completion panel to show the doc, just like other editors

Reply:
We can't do that, because we don't know which completion item is highlighted in the AC panel.

Reply:
Yes, we know up-front whether server can resolve docs through the `resolveProvider` server capability. So if that's not present or set to false then we won't show the hyperlink.

Reply:
But if that's a server capability and not per-completion then I suppose some completion might still return nothing.

Reply:
> we don't know which completion item is highlighted in the AC widget.


I got misled by this 
https://discordapp.com/channels/280102180189634562/280102180189634562/704475737088065566

Reply:
> But if that's a server capability and not per-completion then I suppose some completions might still return nothing.

Yes.